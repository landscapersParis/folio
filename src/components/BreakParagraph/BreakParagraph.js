import React, { Component, PropTypes } from 'react';

export default class BreakParagraph extends Component {
  static propTypes = {
    paragraph: PropTypes.string.isRequired,
    renderPlaceholder: PropTypes.func.isRequired,
    renderLines: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);
    this.spans = [];
    this.state = {
      breakParagraph: false
    };
  }

  componentDidMount() {
    setTimeout(() => {
      this.breakPara();
    });
    window.onresize = () => setTimeout(() => this.breakPara());
    this.listener = () => this.breakPara();
    window.addEventListener('resize', this.listener);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.listener);
  }

  breakPara() {
    let lines = [];
    let tops = [];
    // console.log(this.spans)

    this.spans.forEach((span, i) => {
      const _offsetTop = span.offsetTop;
      let topIndex = tops.indexOf(_offsetTop);
      if (topIndex > -1) {
        if (!lines[topIndex]) lines[topIndex] = [];
      } else {
        topIndex = tops.length;
        tops[tops.length] = _offsetTop;
        if (!lines[topIndex]) lines[topIndex] = [];
      }

      if (span.innerText) {
        lines[topIndex].push(span.innerText.trim());
      }
    });

    lines = lines.map(line => line.join(' '));
    // console.log(lines);

    this.setState({
      breakParagraph: true,
      lines
    });
  };

  setSpanRef(i) {
    return (elem) => this.spans[i] = elem;
  }

  render() {
    const { breakParagraph, lines } = this.state;
    const { paragraph, renderPlaceholder, renderLines } = this.props;

    const spans = paragraph
      .split(' ')
      .map((word, i) => <span
        style={{ pointerEvents: 'none' }}
        key={i}
        ref={this.setSpanRef(i).bind(this)}>{word + ' '}</span>);

    return (
      <div style={{
        width: '100%',
        position: 'relative',
        padding: 0
      }}>
        <div style={{
          opacity: 0,
          position: 'absolute',
          width: '100%'
        }}>
          {renderPlaceholder(spans)}
        </div>
        { breakParagraph && renderLines(lines) }
      </div>
    );
  }
}
