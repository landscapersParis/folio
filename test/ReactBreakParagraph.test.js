import React from 'react';
import BreakParagraph from 'components/BreakParagraph/BreakParagraph';
import { mount } from 'enzyme';
import sinon from 'sinon';
// import renderer from 'react-test-renderer';

jest.useFakeTimers();

describe('react-break-paragraph', function () {
  let wrapper,
    text = 'This is a very long text. This is a very long text. This is a very long text. This is a very long text. This is a very long text. This is a very long text. This is a very long text.',
    testComponent;


  sinon.spy(BreakParagraph.prototype, 'componentDidMount');
  BreakParagraph.prototype.setSpanRef = function(i) {
    return function(elem) {
      this.spans[i] = {
        offsetTop: 0,
        innerText: text.split(' ')[i]
      };
    };
  };

  beforeEach(function() {
    testComponent = (
      <BreakParagraph
        renderPlaceholder={(children) => <h1 className="placeholder">{children}</h1>}
        renderLines={(lines) => <h1>
          {lines.map((line, i) => <span className="line" key={i}>{line}</span>)}
        </h1>}
        paragraph={text}
      />
    );

    wrapper = mount(testComponent);
  });

  afterEach(function() {
    jest.runAllTimers();
    wrapper.unmount();
  });

  it('should call breakPara right after mounted', function() {
    // console.log(BreakParagraph.prototype.breakPara)
    sinon.spy(BreakParagraph.prototype, 'breakPara');
    jest.runAllTimers();
    expect(BreakParagraph.prototype.breakPara.called).toEqual(true);
  });

  it('shoud render', function() {
    expect(wrapper.text().trim()).toEqual(text);
  });

  it('should have at least 1 line', function() {
    jest.runAllTimers();
    expect(wrapper.state().breakParagraph).toBe(true);
    expect(wrapper.find('.line').length).toEqual(1);
  });

  it('should call breakPara when window is resized', function() {
    // sinon.spy(BreakParagraph.prototype, 'breakPara');
    jest.runAllTimers();
    const beforeResizeCount = BreakParagraph.prototype.breakPara.callCount;

    expect(BreakParagraph.prototype.breakPara.callCount).toEqual(beforeResizeCount);
    window.dispatchEvent(new Event('resize'));

    expect(BreakParagraph.prototype.breakPara.callCount).toEqual(beforeResizeCount + 1);
  });
})