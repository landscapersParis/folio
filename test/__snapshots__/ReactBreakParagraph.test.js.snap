exports[`react-break-paragraph shoud render 1`] = `
<div
  style={
    Object {
      "padding": 0,
      "position": "relative",
      "width": "100%",
    }
  }>
  <div
    style={
      Object {
        "opacity": 0,
        "position": "absolute",
        "width": "100%",
      }
    }>
    <h1
      className="placeholder">
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        This 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        is 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        a 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        very 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        long 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        text. 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        This 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        is 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        a 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        very 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        long 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        text. 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        This 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        is 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        a 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        very 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        long 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        text. 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        This 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        is 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        a 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        very 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        long 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        text. 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        This 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        is 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        a 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        very 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        long 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        text. 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        This 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        is 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        a 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        very 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        long 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        text. 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        This 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        is 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        a 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        very 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        long 
      </span>
      <span
        style={
          Object {
            "pointerEvents": "none",
          }
        }>
        text. 
      </span>
    </h1>
  </div>
</div>
`;
